google/cloud/dataproc_spark_connect/session.py

```diff
@@ -669,6 +753,106 @@ def _display_view_session_details_button(self, session_id):
+                logger.debug(f"Deleted session: {session_name}")
+            except NotFound:
+                logger.debug(f"Session already deleted: {session_name}")
+
+        def _wait_for_termination(self, session_name: str, timeout: int = 180):
+            """Wait for a session to finish terminating."""
+            start_time = time.time()
+
+            while time.time() - start_time < timeout:
+                try:
+                    get_request = GetSessionRequest(name=session_name)
+                    session = self.session_controller_client.get_session(
+                        get_request
+                    )
+
+                    if session.state in [
+                        Session.State.TERMINATED,
+                        Session.State.FAILED,
+                    ]:
+                        return
+                    elif session.state != Session.State.TERMINATING:
+                        # Session is in unexpected state
+                        logger.warning(
+                            f"Session {session_name} in unexpected state while waiting for termination: {session.state}"
+                        )
+                        return
+
+                    time.sleep(2)
+                except NotFound:
+                    # Session was deleted
+                    return
+
+            logger.warning(
+                f"Timeout waiting for session {session_name} to terminate"
+            )
+
+        def _get_or_cleanup_session_by_id(
+            self, session_id: str
+        ) -> Optional[Session]:
+            """
+            Get existing session or cleanup terminated/failed sessions.
+
+            Returns:
+                Session if ACTIVE/CREATING, None if not found or cleaned up
+            """
+            session_name = f"projects/{self._project_id}/locations/{self._region}/sessions/{session_id}"
+
+            try:
+                get_request = GetSessionRequest(name=session_name)
+                session = self.session_controller_client.get_session(
+                    get_request
+                )
+
+                logger.debug(
+                    f"Found existing session {session_id} in state: {session.state}"
+                )
+
+                if session.state in [
+                    Session.State.ACTIVE,
+                    Session.State.CREATING,
+                ]:
+                    # Reuse the active session
+                    logger.info(f"Reusing existing session: {session_id}")
+                    return session
+
+                elif session.state in [
+                    Session.State.TERMINATED,
+                    Session.State.FAILED,
+                ]:
+                    # Delete terminated/failed session to free up the ID
+                    logger.info(
+                        f"Cleaning up {session.state.name} session: {session_id}"
+                    )
+                    self._delete_session(session_name)
+                    return None
```

Comment:
I'm not sure whether auto-clean up of sessions is something that we should do - maybe let's remove it for now and delete in the future.